---
title: "Correlation Study and Transcription Factors"
author: "Sofia Cámara Sánchez"
date: "2024-10-03"
output: html_document
---

```{r}
library(mixOmics)
```

```{r}
miRNAs = read.table("miRNAs_data.txt", header = TRUE, row.names = 1, 
                    sep = "\t", stringsAsFactors = TRUE)
RNAseq = read.table("RNAseq_data.txt", header = TRUE, row.names = 1, 
                    sep = "\t", stringsAsFactors = TRUE)
Prot = read.table("proteomic_data.txt", header = TRUE, row.names = 1, 
                  sep = "\t", stringsAsFactors = TRUE)
Met = read.table("metabolomic_data.txt", header = TRUE, row.names = 1, 
                 sep = "\t", stringsAsFactors = TRUE)
SNPs = read.csv("SNPs_gene.csv", header = TRUE, row.names = 1, sep = ",")

Met = log2(Met)
```

```{r}
t_miRNAs = t(miRNAs)
t_RNAseq = t(RNAseq)
t_Prot = t(Prot)
t_Met = t(Met)
t_SNPs = t(SNPs)

t_miRNAs = scale(t_miRNAs, center = TRUE, scale = FALSE)
t_RNAseq = scale(t_RNAseq, center = TRUE, scale = FALSE)
t_Prot = scale(t_Prot, center = TRUE, scale = FALSE)
t_Met = scale(t_Met, center = TRUE, scale = TRUE)
t_SNPs = scale(t_SNPs, center = TRUE, scale = FALSE)
```

```{r}
data = list(rnaseq = t_RNAseq, 
            mirna = t_miRNAs, 
            prot = t_Prot, 
            met = t_Met, 
            snp = t_SNPs)

Y = rep(c("C", "H"), each = 8)
```

```{r}
definitive_design = matrix(0, ncol = length(data), nrow = length(data),
                           dimnames = list(names(data), names(data)))

definitive_design[1,2] = 0.57
definitive_design[1,3] = 0.85
definitive_design[1,4] = 0.44
definitive_design[1,5] = 0.9

definitive_design[2,1] = 0.57
definitive_design[2,3] = 0.62
definitive_design[2,4] = 0.1
definitive_design[2,5] = 0.6

definitive_design[3,1] = 0.85
definitive_design[3,2] = 0.62
definitive_design[3,4] = 0.45
definitive_design[3,5] = 0.95

definitive_design[4,1] = 0.44
definitive_design[4,2] = 0.1
definitive_design[4,3] = 0.45
definitive_design[4,5] = 0.39

definitive_design[5,1] = 0.9
definitive_design[5,2] = 0.6
definitive_design[5,3] = 0.95
definitive_design[5,4] = 0.39
```

```{r}
diablo = block.splsda(X = data, Y = Y, scale = FALSE, ncomp = 2, 
                      design = definitive_design, 
                      near.zero.var = TRUE)
```

```{r}
accumulated_vip_block = function(object, block, ncomp){
  
        if (class(block)!='character'){stop("'block' has to be a character")}
        if (any(class(object) %in% c("block.splsda", "block.spls", "block.plsda"))){
        Y = object$ind.mat }
        else{
          stop( " 'vip_block' is only implemented for the following objects:
                block.splsda or block.plsda ", call.=FALSE) }
  
      
        W = object$loadings[[block]]
        H = ncomp
        q = ncol(Y)
        p = ncol(object$X[[block]])
        VIP = matrix(0, nrow = p, ncol = H)
        
        cor2 = cor(Y, object$variates[[block]], use = "pairwise")^2
        cor2 = as.matrix(cor2, nrow = q)
        
        VIP[, 1] = W[, 1]^2
        
        if (H > 1)  # If there are more than 1 component
        {
            for (h in 2:H)
            {
                if (q == 1)
                {
                    Rd = cor2[, 1:h] 
                    VIP[, h] = Rd %*% t(W[, 1:h]^2) / sum(Rd)
                } else {
                    Rd = apply(cor2[, 1:h], 2, sum)
                    VIP[, h] = Rd %*% t(W[, 1:h]^2) / sum(Rd)
                }
            }
        }
        
        
        VIP = sqrt(p * VIP)
        rownames(VIP) = rownames(W)
        colnames(VIP)= paste0("comp", 1:H)
        
        
        sum_rows = rowSums(VIP)
        ncomp = ncol(VIP)
        accumulated_vip = sum_rows / ncomp

        vip_result = data.frame(Accumulated_VIP = accumulated_vip)
        
        return(invisible(vip_result))
}
```

```{r}
result_rnaseq = accumulated_vip_block(diablo, "rnaseq", ncomp = 1)
result_mirna = accumulated_vip_block(diablo, "mirna", ncomp = 2)
result_prot = accumulated_vip_block(diablo, "prot", ncomp = 1)
result_met = accumulated_vip_block(diablo, "met", ncomp = 1)
result_snp = accumulated_vip_block(diablo, "snp", ncomp = 1)
```

```{r}
vip_ordered = result_rnaseq[order(-result_rnaseq$Accumulated_VIP), , drop = FALSE]
top_100_rnaseq = head(vip_ordered, 100)

vip_ordered = result_prot[order(-result_prot$Accumulated_VIP), , drop = FALSE]
top_100_prot = head(vip_ordered, 100)

vip_ordered = result_snp[order(-result_snp$Accumulated_VIP), , drop = FALSE]
top_100_snp = head(vip_ordered, 100)


top_mirna = result_mirna[result_mirna$Accumulated_VIP > 1, , drop = FALSE]
top_met = result_met[result_met$Accumulated_VIP > 1, , drop = FALSE]
```

```{r}
library(biomaRt)
```

```{r}
rnaseq_vars = rownames(top_100_rnaseq)
mirna_vars = rownames(top_mirna)
prot_vars = rownames(top_100_prot)
met_vars = rownames(top_met)
snp_vars = rownames(top_100_snp)
```

```{r}
ensembl = useEnsembl(biomart = "genes", dataset = "rnorvegicus_gene_ensembl")

genes = getBM(attributes = c('ensembl_gene_id', 'external_gene_name'), 
              filters = 'ensembl_gene_id',
              values = rnaseq_vars, mart = ensembl)

ensembl_df = data.frame(ensembl_gene_id = rnaseq_vars)

association = merge(ensembl_df, genes, by = "ensembl_gene_id", all.x = TRUE, sort = FALSE)

ext_gene_names = ifelse(is.na(association$external_gene_name), 
                        association$ensembl_gene_id,
                        association$external_gene_name)

ext_gene_names = paste0(ext_gene_names, "_r")
definitive_association_rnaseq = setNames(ext_gene_names, rnaseq_vars)
```

```{r}
genes = getBM(attributes = c('ensembl_gene_id', 'external_gene_name'), 
              filters = 'ensembl_gene_id',
              values = snp_vars, mart = ensembl)

ensembl_df = data.frame(ensembl_gene_id = snp_vars)
association = merge(ensembl_df, genes, by = "ensembl_gene_id", all.x = TRUE, sort = FALSE)

ext_gene_names = ifelse(is.na(association$external_gene_name), 
                        association$ensembl_gene_id,
                        association$external_gene_name)

ext_gene_names = paste0(ext_gene_names, "_s")
definitive_association_snp = setNames(ext_gene_names, snp_vars)
```

```{r}
genes = getBM(attributes = c('uniprotswissprot', 'ensembl_gene_id', 'external_gene_name'),
              filters = 'uniprotswissprot', 
              values = prot_vars, mart = ensembl)

ensembl_df = data.frame(protein_id = prot_vars)
association = merge(ensembl_df, genes, by.x = "protein_id", 
                    by.y = "uniprotswissprot", all.x = TRUE,
                    sort = FALSE)

ext_gene_names = ifelse(is.na(association$external_gene_name), 
                        association$protein_id,
                        association$external_gene_name)

ext_gene_names = paste0(ext_gene_names, "_p")

definitive_association_prot = setNames(ext_gene_names, prot_vars)
```

```{r}
short_mirna = gsub("rno-", "", mirna_vars)
```

```{r}
rnaseq_data = RNAseq[rnaseq_vars, ]
mirna_data = miRNAs[mirna_vars, ]
prot_data = Prot[prot_vars, ]
met_data = Met[met_vars, ]
snp_data = SNPs[snp_vars, ]
```


## Transcription Factors

### 1) UniProt

Ahora, para nuestros genes y proteínas, nos interesa saber cuáles de ellos codifican o son **Factores de Trsncripción**. Para ello, nos descargamos una lista de FT de Uniprot para *Rattus Norvegicus*.

```{r}
FT = read.table("uniprotkb_transcription_factors_rattus_2024_10_03.list", 
                header = FALSE)
head(FT)
```

Dimensiones de los FT (tenemos un total de 401) y de los datos que usaremos para generar la red, es decir, 100 son las variables seleccionadas como más relevantes según su VIP.

```{r}
dim(FT)
dim(rnaseq_data)
dim(prot_data)
dim(snp_data)
```

De las variables consideradas como más relevantes, generamos listas que almacenen sus nombres.

```{r}
rnaseq_vars = rownames(top_100_rnaseq)
prot_vars = rownames(top_100_prot)
snp_vars = rownames(top_100_snp)

head(rnaseq_vars)
head(prot_vars)
head(snp_vars)
```

En los scripts anteriores de cambios de ID, se realizaba para conseguir una nomenclatura más informativa pasando de los códigos originales de "ensembl_gene_id" a un "external_gene_name". En este caso, los FT de transcripción son proteínas, por ello estos y mis códigos originales para proteína son los mismos. Vemos que de mis 100 proteínas, 5 son Factores de Transcripción.

```{r}
sum(prot_vars %in% FT$V1)
```

En el caso de los genes, los códigos ya no son los mismos. Voy a coger los nombres de la lista de FT y los voy a pasar a un "ensembl_gene_id" para comparar.

```{r}
ensembl = useEnsembl(biomart = "genes", dataset = "rnorvegicus_gene_ensembl")

genes = getBM(attributes = c('uniprotswissprot', 'ensembl_gene_id'),
              filters = 'uniprotswissprot', 
              values = FT$V1, mart = ensembl)

ensembl_df = data.frame(protein_id = FT$V1)
association = merge(ensembl_df, genes, by.x = "protein_id", 
                    by.y = "uniprotswissprot", all.x = TRUE,
                    sort = FALSE)
head(association)
```

```{r}
sum(rnaseq_vars %in% association$ensembl_gene_id)
sum(snp_vars %in% association$ensembl_gene_id)
```

Vemos los nombres de las proteínas que son FT.

```{r}
protein_FT = prot_vars[prot_vars %in% FT$V1]
protein_FT
```

Vemos los nombres de los genes que codifican para estos FT.

```{r}
protein_gene = association$ensembl_gene_id[association$protein_id %in% prot_vars]
protein_gene
```

Vemos si estos genes se encuentran en RNA-Seq o SNP.

```{r}
any(protein_gene %in% rnaseq_vars)
any(protein_gene %in% snp_vars)
```

Nos centramos únicamente en estas variables de proteínas, realizamos el cambio de ID para la red.

```{r}
ID_change = getBM(attributes = c('uniprotswissprot', 'ensembl_gene_id', 'external_gene_name'),
                  filters = 'uniprotswissprot', 
                  values = protein_FT, mart = ensembl)

ensembl_df = data.frame(protein_id = protein_FT)
association_FT = merge(ensembl_df, ID_change, by.x = "protein_id", 
                    by.y = "uniprotswissprot", all.x = TRUE,
                    sort = FALSE)
head(association_FT)
```

Como podemos ver, con el uso de una lista de FTs obtenidos en Uniprot de *Rattus norvegicus* y revisados, tan solo obtenemos que 5 de nuestras proteínas y ninguno de nuestros genes seleccionados son FTs. Vamos a probar con una anotación en términos GO.

### 2) Gene Ontology (GO)

Conectamos con la base de datos de Ensembl para datos de genes de *Rattus norvegicus* mediante BioMart.

```{r, eval=FALSE}
# Ya lo hemos lanzado antes pero es necesario
ensembl = useEnsembl(biomart = "genes", dataset = "rnorvegicus_gene_ensembl")
```

Organizamos y recogemos la info con `getNM` en la variable "go_annotations". En "rnaseq_vars" están almacenados los nombres de los 100 genes considerados más relevantes por el VIP. Identificamos que el término GO para Factores de Transcripción es el *GO:0003700*.

- **ensembl_gene_id:** Lista de los IDs únicos de Ensembl para cada gen. Identificador estandarizado que se usa en las bases de datos genómicas para referirse de manera única a cada gen.
- **external_gene_name:** Nombre externo del gen, que es el nombre más comúnmente utilizado en la literatura científica.
- **go_id:** ID del término Gene Ontology (GO) asociado con el gen.
- **name_1006:** Nombre o descripción del término GO asociado al gen. Por ejemplo, si el go_id fuese "GO:0003700" (el que buscamos), el name_1006 sería "DNA binding transcription factor activity".

```{r}
go_annotations = getBM(
  attributes = c("ensembl_gene_id", "external_gene_name", "go_id", "name_1006"),
  filters = "ensembl_gene_id", 
  values = rnaseq_vars,
  mart = ensembl
)

head(go_annotations)
```

Una vez tenemos la tabla montada para todos nuestros genes seleccionados, vamos a filtrar por el término GO (go_id) y nos vamos a quedar solamente con aquellos que tienen el "GO:0003700". Solamente encontramos uno.

```{r}
go_TFactors = go_annotations[go_annotations$go_id == "GO:0003700", ]
head(go_TFactors)
```

Repetimos lo mismo para SNP y Proteomics.

```{r}
go_annotations2 = getBM(
  attributes = c("ensembl_gene_id", "external_gene_name", "go_id", "name_1006"),
  filters = "ensembl_gene_id", 
  values = snp_vars,
  mart = ensembl
)

head(go_annotations2)
```

```{r}
go_TFactors2 = go_annotations2[go_annotations2$go_id == "GO:0003700", ]
head(go_TFactors2)
```

```{r}
prot_annotations = getBM(
  attributes = c("uniprotswissprot", "ensembl_gene_id", "go_id", "name_1006"),
  filters = "uniprotswissprot", 
  values = prot_vars,  
  mart = ensembl
)

head(prot_annotations)
```

```{r}
prot_TFactors = prot_annotations[prot_annotations$go_id == "GO:0003700", ]
head(prot_TFactors)
```


```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

